incubator/hnc/internal/validators/object.go

```diff
@@ -80,37 +97,44 @@ func (o *Object) isPropagateType(gvk metav1.GroupVersionKind) bool {
-func (o *Object) handle(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured, oldInst *unstructured.Unstructured) admission.Response {
-	// We want to ignore validation for objects in the exclusion list.
-	if config.EX[inst.GetNamespace()] {
-		return allow("")
+func (o *Object) handle(ctx context.Context, log logr.Logger, op admissionv1beta1.Operation, inst, oldInst *unstructured.Unstructured) admission.Response {
+	// Find out if the object was/is inherited, and where it's inherited from.
+	oldSource, oldInherited := metadata.GetLabel(oldInst, api.LabelInheritedFrom)
+	newSource, newInherited := metadata.GetLabel(inst, api.LabelInheritedFrom)
+
+	// If the object wasn't and isn't inherited, it's none of our business.
+	if !oldInherited && !newInherited {
+		return allow("source object")
 	}
 
-	// Prevent users from changing the InheritedFrom label
-	oldValue, oldExists := metadata.GetLabel(oldInst, api.LabelInheritedFrom)
-	newValue, newExists := metadata.GetLabel(inst, api.LabelInheritedFrom)
+	switch op {
+	case admissionv1beta1.Create:
+		return deny(metav1.StatusReasonForbidden, "Cannot create objects with the label \""+api.LabelInheritedFrom+"\"")
+
+	case admissionv1beta1.Delete:
+		return deny(metav1.StatusReasonForbidden, "Cannot delete object propagated from namespace \""+oldSource+"\"")
 
-	// If old object holds the label but the new one doesn't, reject it. Vice versa.
-	if oldExists != newExists {
-		verb := "add"
-		if !newExists {
-			verb = "remove"
+	case admissionv1beta1.Update:
+		// If the values have changed, that's an illegal modification. This includes if the label is
+		// added or deleted.
+		if newSource != oldSource {
+			return deny(metav1.StatusReasonForbidden, "Cannot modify the label \""+api.LabelInheritedFrom+"\"")
+		}
+
+		// If the existing object has an inheritedFrom label, it's a propagated object. Any user changes
+		// should be rejected. Note that object.Canonical does *not* compare any HNC labels or
+		// annotations.
+		if !reflect.DeepEqual(object.Canonical(inst), object.Canonical(oldInst)) {
+			return deny(metav1.StatusReasonForbidden,
+				"Cannot modify object propagated from namespace \""+oldSource+"\"")
 		}
-		return deny(metav1.StatusReasonForbidden, "Users should not "+verb+" the label "+api.LabelInheritedFrom)
-	}
-	// If both old and new objects hold the label but with different values, reject it.
-	if newExists && newValue != oldValue {
-		return deny(metav1.StatusReasonForbidden, "Users should not change the value of label "+api.LabelInheritedFrom)
-	}
 
-	// If the existing object has an inheritedFrom label, it's a propagated object.
-	// Any user changes should be rejected.
-	if newExists && !reflect.DeepEqual(object.Canonical(inst), object.Canonical(oldInst)) {
-		return deny(metav1.StatusReasonForbidden,
-			"Illegal modification to an object propagated by the Hierarchical Namespace Controller")
+		return allow("all modifications allowed")
```

Comment:
What does the modifications imply here?

Reply:
Oh it's still in the case of `update`, but the only possibility is `DeepEqual` here, right?

Reply:
Yup, updated comment, PTAL